**Goal**: to understand the features where are given in this dataset.

In [12]:
import pandas as pd
import os

In [13]:
['data/' + path for path in os.listdir('data')]

['data/heart_disease_x_features.csv',
 'data/heart_disease_y_labels.csv',
 'data/project11_heart disease dataset.zip']

In [15]:
def read(path: str):
    return pd.read_csv(path)

In [16]:
x = read('data/heart_disease_x_features.csv')
y = read('data/heart_disease_y_labels.csv')

First, I want to make sure that both datasets are the same size

In [17]:
x.shape

(303, 13)

In [18]:
y.shape

(303, 1)

## Y Label
Now, let's explore the y labels, this should be self explanatory i.e if heart disease or not

In [22]:
y.head()

,label
0,0
1,1
2,1
3,0
4,0


okay since there is no label name, further confirmation is needed but for now, we can assume that a value of 0 means FALSE meaning not heart disease and a value of 1 means TRUE meaning heart disease

## X Label
Now, let's explore the x labels including what they mean. As I research I will write down my findings and sources

In [23]:
x.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0


In [26]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
dtypes: float64(3), int64(10)
memory usage: 30.9 KB


## Feature Dictionary
link: https://archive.ics.uci.edu/dataset/45/heart+disease

### Quick Guide
- **age**: the age of a person in years
- **sex**: the sex of a person (1 = male; 0 = female)
- **cp**: chest pain type:
    - 1 = typical angina
    - 2 = atypical angine
    - 3 = non-anginal pain
    - 4 = asymptomatic
- **trestbps**: resting blood pressure (in mm Hg on admission to the hospital)
- **chol**: serum cholestoral in mg/dl
- **fbs**: if fasting blood sugar > 120 mg/dl: (1 = true; 0 = false)
- **restecg**: resting electrocardiographic results:
    - 0 = normal
    - 1 = having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    - 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria
- **thalach**: maximum heart rate achieved
- **exang**: exercise induced angina (1 = yes; 0 = no)
- **oldpeak**: ST depression induced by exercise relative to rest
- **slope**: the slope of the peak exercise ST segment
    - 1 = upsloping
    - 2 = flat
    - 3 = downsloping
- **ca**: number of major vessels (0-3) colored by flourosopy
- **thal**: 3 = normal; 6 = fixed defect; 7 = reversable defect

### Extended Guide


#### `age` - Age in Years

Age is one of the most powerful independent risk factors for coronary heart disease in this dataset. Data drawn from roughly 1 million individuals across 80 cohort studies found that for each 5-year increase in age starting from 40, the 10-year risk of fatal and nonfatal heart attack or coronary death increased independently by 43% in men and 67% in women. This makes age more impactful on absolute risk than even a 20 mmHg rise in blood pressure. The dataset spans a wide adult age range, so this column is useful both as a standalone filter and as context for interpreting other measurements - some ECG readings and stress test results mean different things at different ages.

**Common questions:**

- *Why does age matter so much?* Years of accumulated exposure to blood pressure, cholesterol, inflammation, and oxidative stress cause progressive arterial damage. The heart and vasculature undergo structural and functional changes over time that independently increase susceptibility to disease, even when other risk factors are held constant.
- *Is this the person's age at diagnosis?* Yes - it reflects the patient's age at hospital admission during the study.

> **Source:** Rana, J.S. et al. (2024). *Impact of Aging on Cardiovascular Diseases: From Chronological Observation to Biological Insights.* JACC: Asia. https://www.jacc.org/doi/10.1016/j.jacasi.2024.02.002

---

#### `sex` - Biological Sex (1 = Male, 0 = Female)

Sex is included because it significantly modifies both the risk and the presentation of heart disease. Among middle-aged adults, coronary heart disease is 2 to 5 times more common in men than women. However, this gap narrows substantially after menopause, and women often present with less "classic" symptoms, which can make diagnosis harder. Sex also interacts with other features in the dataset: the chest pain classification system (see `cp`) was developed primarily around male presentations and is documented to perform less reliably in women.

**Common questions:**

- *Does this reflect gender identity?* No - this is biological sex as recorded at the time of the study in the 1980s–90s. The dataset does not capture gender identity.
- *Why might women be underrepresented in heart disease datasets?* Women develop coronary artery disease on average 7–10 years later than men, so studies weighted toward middle age will capture fewer female cases. Additionally, atypical symptom presentation in women has historically led to under-referral for diagnostic workup.

> **Source:** Jousilahti, P. et al. (1999). *Sex, Age, Cardiovascular Risk Factors, and Coronary Heart Disease.* Circulation. https://www.ahajournals.org/doi/10.1161/01.CIR.99.9.1165

---

#### `cp` - Chest Pain Type

This is a clinical classification of how a patient describes their chest discomfort. It has four categories ranging from "textbook" heart-related pain to no pain at all.

**Value 1 - Typical Angina**

Typical (classic) angina meets all three of the following criteria: (1) substernal chest discomfort of characteristic quality and duration, (2) provoked by exertion or emotional stress, and (3) relieved by rest and/or nitrate medications like nitroglycerin within minutes. All three must be present.

**Value 2 - Atypical Angina**

Atypical angina applies when only two of the three criteria above are present. The pain may be triggered by exertion and relieved by rest, but located somewhere other than directly beneath the sternum - such as the jaw or left arm. It reduces the prior probability that coronary artery disease is the cause, but does not rule it out.

**Value 3 - Non-Anginal Pain**

Chest pain that meets one or none of the classic angina criteria. It is considered unlikely to be cardiac in origin based on its characteristics alone. However, research on over 3,000 outpatients found that a meaningful proportion of patients classified as non-anginal still had functionally positive stress tests, so the label does not eliminate cardiac risk entirely.

**Value 4 - Asymptomatic**

The patient reports no chest pain at all. Importantly, in this dataset "asymptomatic" does not mean the patient is healthy - many patients with confirmed coronary artery disease in this cohort have no chest pain. This is sometimes called "silent ischemia."

**Common questions:**

- *Why is someone with no chest pain in a heart disease dataset?* Patients were referred for diagnostic workup for many reasons beyond chest pain - abnormal ECGs, elevated risk factor profiles, or related symptoms like shortness of breath.
- *Is this classification still used today?* Its use is debated. Modern cardiology guidelines have moved toward multi-dimensional risk scoring, partly because the typical/atypical/non-anginal framework was developed primarily on male patients and has documented limitations in women, elderly patients, and those with diabetes.

> **Sources:**
> - Prati, F. et al. (2023). *Typical and atypical chest pain: an equivocal and dangerous definition.* European Heart Journal. https://academic.oup.com/eurheartj/article/44/Supplement_2/ehad655.2410/7392415
> - Sionis, A. et al. (2020). *Typical angina, atypical angina, and atypical chest pain: is it time to change this terminology?* Revista Española de Cardiología. https://www.revespcardiol.org/en-typical-angina-atypical-angina-atypical-articulo-S1885585720304692

---

#### `trestbps` - Resting Blood Pressure (mm Hg)

This is the patient's blood pressure at rest, measured in millimeters of mercury (mm Hg), recorded upon hospital admission. The dataset records a single value corresponding to systolic pressure - the pressure in the arteries when the heart beats (the larger number in a reading like 130/85). Normal resting systolic blood pressure is generally below 120 mm Hg. Stage 1 hypertension begins at 130 mm Hg, and stage 2 at 140 mm Hg and above per current ACC/AHA guidelines.

Chronically elevated blood pressure forces the heart to work harder, which stiffens and damages arterial walls over time, accelerating atherosclerosis and increasing the risk of heart attack and stroke.

**Common questions:**

- *Why is it measured "on admission"?* Hospital admission provides a standardized measurement point, but it can be slightly elevated due to the stress of being admitted ("white coat hypertension"), so it may not perfectly reflect a patient's typical resting blood pressure.
- *Is this systolic or diastolic?* Based on the dataset documentation, this records systolic blood pressure.

> **Source:** Whelton, P.K. et al. (2018). *2017 ACC/AHA Guideline for the Prevention, Detection, Evaluation, and Management of High Blood Pressure in Adults.* Journal of the American College of Cardiology. https://www.jacc.org/doi/10.1016/j.jacc.2017.11.006

---

#### `chol` - Serum Cholesterol (mg/dL)

This is the total amount of cholesterol measured in the patient's blood (serum), in milligrams per deciliter. Total cholesterol combines LDL ("bad") cholesterol, HDL ("good") cholesterol, and VLDL. Elevated LDL contributes to the buildup of fatty plaques inside artery walls (atherosclerosis), which can narrow or block coronary arteries and trigger heart attacks. The 2018 AHA/ACC guidelines classify total cholesterol below 200 mg/dL as desirable, 200–239 mg/dL as borderline high, and 240 mg/dL and above as high.

**Common questions:**

- *Why is it called "serum" cholesterol?* Serum is the liquid part of blood after clotting factors are removed. Measuring cholesterol in serum is the standard clinical method.
- *Does this distinguish between LDL and HDL?* No - only total cholesterol is recorded. High HDL is actually cardioprotective, so two patients with identical total cholesterol can have very different risk profiles. This is a known limitation of the dataset.
- *Are there unusually high values?* Yes - some values exceed 500 mg/dL, which may reflect data entry anomalies or severe familial hypercholesterolemia.

> **Source:** Grundy, S.M. et al. (2018). *2018 AHA/ACC Guideline on the Management of Blood Cholesterol.* Circulation. https://www.ahajournals.org/doi/10.1161/CIR.0000000000000625

---

#### `fbs` - Fasting Blood Sugar > 120 mg/dL (1 = True, 0 = False)

This is a binary flag indicating whether a patient's blood glucose, measured after an overnight fast, exceeded 120 mg/dL. It serves as a proxy for diabetic or pre-diabetic status. Diabetes is a well-established independent risk factor for coronary heart disease. Research has found that coronary heart disease risk increases significantly at fasting blood glucose above 120 mg/dL, even in patients not yet meeting formal diabetes criteria - with one large angiography study reporting an odds ratio of 2.7 for coronary heart disease at this threshold.

The 120 mg/dL cutoff used here is slightly higher than the current ADA definition of impaired fasting glucose (which begins at 100 mg/dL). This reflects the diagnostic standards in use at the time the dataset was collected in the 1980s.

**Common questions:**

- *Why 120 mg/dL and not 126 mg/dL (the current diabetes cutoff)?* The dataset predates current ADA thresholds. The 126 mg/dL cutoff for diabetes diagnosis was established in 1997.
- *Does a value of 0 mean the patient is not diabetic?* Not necessarily - a patient could have a fasting glucose between 100 and 120 mg/dL (pre-diabetic range by today's standards) and still be coded 0.

> **Sources:**
> - Schnell, O. et al. (2011). *A progressive increase in cardiovascular risk assessed by coronary angiography in non-diabetic patients at sub-diabetic glucose levels.* GMS German Medical Science. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3142488/
> - Dauriz, M. et al. (2021). *Prediabetes and risk of heart failure: the link grows stronger.* Cardiovascular Diabetology. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8146677/

---

#### `restecg` - Resting Electrocardiographic Results

An electrocardiogram (ECG or EKG) records the electrical signals produced by the heart at each beat. This feature describes what the patient's resting ECG shows, with three possible categories.

**Value 0 - Normal**

No notable abnormalities detected in the heart's resting electrical pattern.

**Value 1 - ST-T Wave Abnormality**

The ST segment and T wave on the ECG correspond to the heart muscle recovering after each beat (repolarization). Abnormalities here - including T wave inversions or ST segment deviations greater than 0.05 mV - can indicate that part of the heart muscle is not receiving enough blood (ischemia) or is under chronic strain. These changes can appear at rest when the heart is persistently stressed, even without a current episode of pain.

**Value 2 - Left Ventricular Hypertrophy (by Estes' Criteria)**

Left ventricular hypertrophy (LVH) means the muscular wall of the heart's main pumping chamber has thickened abnormally - typically in response to sustained high blood pressure or other conditions that force the heart to pump harder over years. On an ECG, this shows up as unusually large electrical signals and characteristic repolarization changes.

"Estes' criteria" refers to the Romhilt-Estes point score - a system that assigns points across multiple ECG features including voltage amplitude, ST-segment pattern, left atrial involvement, axis deviation, and QRS duration. A score of ≥5 is considered definitive for LVH; ≥4 is considered probable. Studies comparing Romhilt-Estes to cardiac MRI have found it offers the most favorable combination of sensitivity and specificity among common ECG-based LVH criteria, with one study reporting sensitivity of 86% and specificity of 81% at the 4-point threshold.

**Common questions:**

- *Is ECG the best way to detect LVH?* No - echocardiography and cardiac MRI are more accurate. However, ECG is inexpensive and widely available, which is why it was used here.
- *Can LVH-related ST changes look like a heart attack on the ECG?* Yes - repolarization abnormalities from LVH can closely mimic ischemic ST changes, which is a well-documented clinical challenge.

> **Sources:**
> - Mori, H. et al. (2022). *The Use of Romhilt-Estes Criteria in the Presumptive Electrocardiographic Diagnosis of Left Ventricular Hypertrophy.* Journal of Clinical Medicine. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9470005/
> - Wachtell, K. et al. (2009). *Electrocardiographic diagnosis of left ventricular hypertrophy in aortic valve disease: evaluation of ECG criteria by cardiovascular magnetic resonance.* BMC Cardiovascular Disorders. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2696426/

---

#### `thalach` - Maximum Heart Rate Achieved

This is the highest heart rate (in beats per minute) the patient reached during a treadmill exercise stress test. For a healthy individual, maximum predicted heart rate is commonly estimated as 220 minus age. Achieving a high proportion of this age-predicted maximum is generally associated with good cardiovascular fitness. Failing to raise heart rate adequately during exercise - called chronotropic incompetence - has been shown to be an independent predictor of overall mortality, even after adjusting for other known cardiac risk factors. In this dataset, a lower-than-expected maximum heart rate is associated with higher heart disease risk.

**Common questions:**

- *Is a higher max heart rate always better?* In the context of an exercise stress test, reaching a higher percentage of age-predicted maximum is generally a positive sign. An abnormally high resting heart rate is a separate concern and is not captured by this variable.
- *What counts as an adequate stress test?* Reaching at least 85% of age-predicted maximum heart rate is generally required for the test to be considered diagnostically valid.

> **Source:** Lauer, M.S. et al. (1999). *Impaired chronotropic response to exercise stress testing as a predictor of mortality.* JAMA. https://jamanetwork.com/journals/jama/fullarticle/191571

---

#### `exang` - Exercise-Induced Angina (1 = Yes, 0 = No)

This records whether the patient experienced chest pain (angina) specifically during the exercise stress test. Angina triggered by physical exertion - and typically relieved by rest - is a strong indication that the heart's blood supply cannot keep pace with its oxygen demand during activity. Increased exertion unmasks coronary artery insufficiency that may not be apparent at rest, making this one of the more direct indicators of functionally significant narrowing.

**Common questions:**

- *If exang = 0, does that mean the heart is healthy?* Not necessarily. A significant number of patients with confirmed coronary artery disease do not feel pain during ischemia ("silent ischemia"). ECG changes during exercise - captured separately by `oldpeak` and `slope` - may still reveal a problem even without reported pain.
- *What does exercise-induced angina feel like?* It is classically described as pressure, squeezing, heaviness, or tightness in the chest, sometimes radiating to the left arm or jaw. It typically resolves within minutes of stopping exercise.

> **Source:** Gibbons, R.J. et al. (2002). *ACC/AHA Guidelines for Exercise Testing.* Journal of the American College of Cardiology. https://www.jacc.org/doi/10.1016/S0735-1097(02)81963-5

---

#### `oldpeak` - ST Depression Induced by Exercise Relative to Rest

During an exercise stress test the ECG is monitored continuously. This feature measures how much the ST segment drops (depresses) during peak exercise compared to the resting baseline, recorded in millivolts. ST segment depression during exercise is a well-established marker of myocardial ischemia - the heart muscle not receiving enough blood when it demands more oxygen.

A value of 0 means no ST depression occurred. Higher values indicate greater depression. An exercise stress test generally requires ST depression of at least 2 mm to significantly indicate reversible ischemia, though 1 mm of horizontal or downsloping depression is widely used as the threshold for a positive test. Larger depressions occurring at lower workloads indicate more severe disease.

**Common questions:**

- *Why does the ST segment dip during exercise?* When a coronary artery is narrowed, it may supply adequate blood at rest but cannot meet the heart's increased oxygen demand during exertion. This shortfall disrupts the electrical recovery process of the heart muscle, pushing the ST segment downward on the ECG.
- *Is a higher oldpeak value always worse?* Generally yes - but the shape of the depression (captured by `slope`) also matters significantly for interpretation.

> **Sources:**
> - Gibbons, R.J. et al. (2002). *ACC/AHA Guidelines for Exercise Testing.* Journal of the American College of Cardiology. https://www.jacc.org/doi/10.1016/S0735-1097(02)81963-5
> - Narula, A. et al. (2021). *Significance of 'recovery ST-segment depression' in exercise stress test.* Indian Heart Journal. https://pmc.ncbi.nlm.nih.gov/articles/PMC8642660/

---

#### `slope` - Slope of the Peak Exercise ST Segment

This describes the shape (angle) of the ST segment at peak exercise. It refines what `oldpeak` tells us by capturing not just how much the ST segment dropped, but the direction in which it moved. The same millimeter of ST depression carries different prognostic weight depending on its shape.

**Value 1 - Upsloping**

The ST segment rises as it moves forward after the J-point. Upsloping ST depression is generally considered the least concerning pattern. A long-term study of over 2,000 healthy middle-aged men found that a rapidly upsloping ST segment was associated with a 30% *reduced* risk of dying from coronary heart disease compared to a normal ST segment.

**Value 2 - Flat (Horizontal)**

The ST segment runs horizontally after the J-point. Horizontal depression of ≥1 mm is the classic positive finding in a stress test and is strongly associated with myocardial ischemia and future coronary events.

**Value 3 - Downsloping**

The ST segment continues to fall as time proceeds. Downsloping depression is the most concerning pattern - it is associated with the most extensive ischemia and the highest severity of coronary artery disease. Among patients with downsloping ST depression, studies have found that 62% had a large reversible ischemic area on nuclear imaging, compared to 25% for horizontal and 20% for upsloping patterns.

**Common questions:**

- *Can the same patient show different slopes across different leads?* Yes. This dataset captures the slope at peak exercise in the lead showing the most significant change.
- *Is upsloping ST depression always benign?* Not always. Slowly upsloping depression - particularly with a slope less than 1 mV/second - can still carry elevated risk, especially in symptomatic patients.

> **Sources:**
> - Hodnesdal, C. et al. (2013). *Rapidly upsloping ST-segment on exercise ECG: a marker of reduced coronary heart disease mortality risk.* European Journal of Preventive Cardiology. https://journals.sagepub.com/doi/10.1177/2047487312444370
> - Safi, M. et al. (2019). *The Value of Upsloping ST Depression in Diagnosing Myocardial Ischemia.* Annals of Noninvasive Electrocardiology. https://pmc.ncbi.nlm.nih.gov/articles/PMC6932726/

---

#### `ca` - Number of Major Vessels Colored by Fluoroscopy (0–3)

This value comes from coronary fluoroscopy - an X-ray imaging technique where a contrast dye is injected into the coronary arteries to make them visible. The result records how many of the three major coronary arteries show significant narrowing or calcification visible under fluoroscopy. The range is 0 (no major vessels visibly affected) to 3 (all three major arteries affected).

The three major coronary vessels are the left anterior descending artery (LAD), the left circumflex artery (LCX), and the right coronary artery (RCA). The number of visibly diseased vessels is one of the strongest direct predictors of heart disease severity in this dataset.

**Common questions:**

- *What is the difference between fluoroscopy here and a full coronary angiogram?* A full angiogram provides detailed visualization and precise quantification of the degree of narrowing (stenosis percentage). Fluoroscopy in this context captures visible calcification and gross obstruction. Both use contrast dye, but angiography is more detailed and quantitative.
- *Does ca = 0 mean the heart is definitely healthy?* No - coronary artery disease can exist without enough calcification to appear on fluoroscopy. A value of 0 indicates no major visible disease, not confirmed absence of all disease.
- *Why does the range stop at 3?* Because there are three main coronary arteries. A value of 3 means all three are visibly diseased (three-vessel disease), which is associated with the worst prognosis and is often an indication for bypass surgery.

> **Source:** Detrano, R. et al. (1989). *International application of a new probability algorithm for the diagnosis of coronary artery disease.* American Journal of Cardiology. (Original dataset paper - linked via UCI repository: https://archive.ics.uci.edu/dataset/45/heart+disease)

---

#### `thal` - Thallium Stress Test Result

This feature comes from a **thallium nuclear stress test** - a type of medical imaging that uses a small radioactive tracer (thallium-201) injected into the bloodstream to produce pictures of blood flow through the heart muscle. A special gamma camera detects the tracer: living heart muscle cells that are receiving blood absorb it and appear bright, while areas with reduced or absent blood flow appear as dark "cold spots." Images are taken both at rest and during exercise stress, and the two sets are compared.

**Value 3 - Normal**

Blood flow through the heart muscle is normal at both rest and during stress. No significant cold spots appear on either scan.

**Value 6 - Fixed Defect**

A cold spot is visible on both the resting and the stress images - the defect does not change between the two states (it is "fixed"). A persistent defect signifies scarred myocardium - dead tissue from a prior heart attack. Because the tissue is no longer viable, restoring blood flow through revascularization is unlikely to improve function in that region.

**Value 7 - Reversible Defect**

A cold spot appears during exercise stress but fills in (resolves) on the resting images. This pattern - also described as "fill-in" - indicates living but ischemic heart muscle: tissue that receives adequate blood at rest but not enough during exertion. This is the hallmark finding of functionally significant coronary artery disease with salvageable tissue, and it is the pattern that most strongly supports pursuing revascularization (stenting or bypass surgery), as the affected muscle has a high likelihood of recovering function.

**Common questions:**

- *Is thallium radioactive? Is the test safe?* A tiny, medically controlled dose is used. The radiation exposure is comparable to other standard medical imaging tests and is considered safe for most patients.
- *Why are the coded values 3, 6, and 7 instead of 1, 2, 3?* This is an artifact of the original dataset's coding scheme and carries no clinical significance - the numbers are category labels, not quantities.
- *What is the practical difference between a fixed and reversible defect for treatment decisions?* A reversible defect strongly supports revascularization because the tissue is alive and at risk. A fixed defect suggests a completed infarction, meaning the tissue is already scarred and revascularization is less likely to improve function there.

> **Sources:**
> - Dilsizian, V. & Bonow, R.O. (2008). *Nuclear Cardiac Stress Testing in the Era of Molecular Medicine.* Journal of Nuclear Medicine. https://jnm.snmjournals.org/content/49/3/399
> - Syed, I.S. et al. (2007). *Adenosine SPECT Thallium Imaging.* StatPearls, NCBI Bookshelf. https://www.ncbi.nlm.nih.gov/books/NBK537070/


In [32]:
# merging the x and y so we have a unified dataset
combined_df = pd.concat([x, y], axis=1)
combined_df.to_csv('data/heart_disease_combined.csv', index=False)